Components/Motors/VESC.cs

```diff
@@ -0,0 +1,190 @@
+        /// <param name="Speed"> Speed from -1.0 to 1.0 </param>
+        private void SetSpeedDirectly(float Speed)
+        {
+            if (Speed > this.MaxSpeed) { Speed = this.MaxSpeed; }
+            if (-Speed > this.MaxSpeed) { Speed = -this.MaxSpeed; }
+            if (this.Stopped) { Speed = 0; }
+            UARTBus.Write(ConstructPacket(Speed));
+        }
+
+
+        /// <summary>
+        /// Generates the packet for the motor controller:
+        /// One Start byte (value 2 for short packets and 3 for long packets)
+        /// One or two bytes specifying the packet length
+        /// The payload of the packet
+        /// Two bytes with a CRC checksum on the payload
+        /// One stop byte (value 3)
+        /// </summary>
+        /// <param name="Speed"> Speed from -1.0 to 1.0 </param>
+        private byte[] ConstructPacket(float Speed)
+        {
+            List<byte> packet = new List<byte>();
+            List<byte> payload = new List<byte>();
+            byte ShortPacket = 2;
+
+            packet.Add(ShortPacket); // Start byte
+            
+            if (CANForwardId >= 0)
+            {
+                payload.Add((byte) PacketId.FORWARD_CAN);
+                payload.Add((byte) CANForwardId);
+            }
+
+            payload.Add((byte) PacketId.SET_DUTY);
+            // Duty Cycle (100000 mysterious magic number from https://github.com/VTAstrobotics/VESC_BBB_UART/blob/master/bldc_interface.c)
+            payload.AddRange(UtilData.ToBytes((Int32)(Speed * 100000.0)));
+
+            packet.Add((byte) payload.Count); // Length of payload
+            packet.AddRange(payload); // Payload
+
+            ushort checksum = UtilData.CRC16(payload.ToArray());
+            packet.AddRange(UtilData.ToBytes(checksum)); // Checksum
+
+            packet.Add(3); // Stop byte
+
+            return packet.ToArray();
+        }
+
+        #region enums
+        private enum PacketId : byte
+        {
+            FW_VERSION,
+            JUMP_TO_BOOTLOADER,
+            ERASE_NEW_APP,
+            WRITE_NEW_APP_DATA,
+            GET_VALUES,
+            SET_DUTY,
+            SET_CURRENT,
+            SET_CURRENT_BRAKE,
+            SET_RPM,
+            SET_POS,
+            SET_DETECT,
+            SET_SERVO_POS,
+            SET_MCCONF,
+            GET_MCCONF,
+            GET_MCCONF_DEFAULT,
+            SET_APPCONF,
+            GET_APPCONF,
+            GET_APPCONF_DEFAULT,
+            SAMPLE_PRINT,
+            TERMINAL_CMD,
+            PRINT,
+            ROTOR_POSITION,
+            EXPERIMENT_SAMPLE,
+            DETECT_MOTOR_PARAM,
+            DETECT_MOTOR_R_L,
+            DETECT_MOTOR_FLUX_LINKAGE,
+            DETECT_ENCODER,
+            DETECT_HALL_FOC,
+            REBOOT,
+            ALIVE,
+            GET_DECODED_PPM,
+            GET_DECODED_ADC,
+            GET_DECODED_CHUK,
+            FORWARD_CAN,
+            SET_CHUCK_DATA,
+            CUSTOM_APP_DATA
+        }

```

Comment:
You define all of these possible packets, but only support one single type. Will there be applications where other packets may be necessary or useful? If so, we should implement them.

Reply:
Getting there!